File: carddemo-java/src/main/java/com/carddemo/model/DisclosureGroup.java
package com.carddemo.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Table;
import java.math.BigDecimal;

/**
 * Disclosure Group entity - mapped from COBOL copybook CVTRA02Y.cpy (DISCGRP-RECORD).
 * Original VSAM dataset: AWS.M2.CARDDEMO.DISCGRP.VSAM.KSDS (RECLN 50).
 * Composite key: DIS-ACCT-GROUP-ID + DIS-TRAN-TYPE-CD + DIS-TRAN-CAT-CD.
 * Used by batch program CBACT04C for interest rate lookups.
 */
@Entity
@Table(name = "disclosure_groups")
@IdClass(DisclosureGroupId.class)
public class DisclosureGroup {

    @Id
    @Column(name = "dis_acct_group_id", length = 10)
    private String disAcctGroupId;

    @Id
    @Column(name = "dis_tran_type_cd", length = 2)
    private String disTranTypeCd;

    @Id
    @Column(name = "dis_tran_cat_cd")
    private Integer disTranCatCd;

    @Column(name = "dis_int_rate", precision = 7, scale = 4, nullable = false)
    private BigDecimal disIntRate = BigDecimal.ZERO;

    @Column(name = "dis_fee_amt", precision = 9, scale = 2, nullable = false)
    private BigDecimal disFeeAmt = BigDecimal.ZERO;

    public DisclosureGroup() {
    }

    public String getDisAcctGroupId() {
        return disAcctGroupId;
    }

    public void setDisAcctGroupId(String disAcctGroupId) {
        this.disAcctGroupId = disAcctGroupId;
    }

    public String getDisTranTypeCd() {
        return disTranTypeCd;
    }

    public void setDisTranTypeCd(String disTranTypeCd) {
        this.disTranTypeCd = disTranTypeCd;
    }

    public Integer getDisTranCatCd() {
        return disTranCatCd;
    }

    public void setDisTranCatCd(Integer disTranCatCd) {
        this.disTranCatCd = disTranCatCd;
    }

    public BigDecimal getDisIntRate() {
        return disIntRate;
    }

    public void setDisIntRate(BigDecimal disIntRate) {
        this.disIntRate = disIntRate;
    }

    public BigDecimal getDisFeeAmt() {
        return disFeeAmt;
    }

    public void setDisFeeAmt(BigDecimal disFeeAmt) {
        this.disFeeAmt = disFeeAmt;
    }
}

File: carddemo/app/models/disclosure_group.py
"""Disclosure group model — from CVTRA02Y.cpy (50-byte VSAM KSDS).

COBOL record layout:
  05 DIS-GROUP-KEY
     10 DIS-ACCT-GROUP-ID  PIC X(10)   → String(10)
     10 DIS-TRAN-TYPE-CD   PIC X(02)   → String(2)
     10 DIS-TRAN-CAT-CD    PIC 9(04)   → Integer
  05 DIS-INT-RATE          PIC S9(04)V99 → Numeric(6,2)
  05 FILLER                PIC X(28)   → not stored

Used by CBACT04C.cbl to look up interest rates for each transaction category.
Composite primary key: (group_id, tran_type_cd, tran_cat_cd)
"""

from sqlalchemy import Integer, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class DisclosureGroup(Base):
    __tablename__ = "disclosure_groups"

    group_id: Mapped[str] = mapped_column(String(10), primary_key=True)
    tran_type_cd: Mapped[str] = mapped_column(String(2), primary_key=True)
    tran_cat_cd: Mapped[int] = mapped_column(Integer, primary_key=True)
    interest_rate: Mapped[object] = mapped_column(Numeric(6, 2), default=0)
